File: fruit-machine/fruit_machine.py

import random


class Player(object):
    """Player has a fund of cash (integer) to play machines with"""
    def __init__(self, wallet_fund):
        self.wallet_fund = wallet_fund

    def gamble(self, machine):
        if (self.wallet_fund - machine.play_cost) > 0:
            self.wallet_fund -= machine.play_cost
            machine.run()


class Machine(object):
    """Each machine has an initial bank and an cost (integer) for each turn"""
    def __init__(self, machine_bank, play_cost):
        self.machine_bank = machine_bank
        self.play_cost = play_cost
        self.play_credit = 0

    def run(self):
        self.credit_or_cash()
        result = self.roll_slots()
        print result
        prize = self.determine_prize(result)
        print 'WINNINGS: %i' % prize

    def credit_or_cash(self):
        if self.play_credit > 0:
            self.play_credit -= 1
        else:
            self.machine_bank += self.play_cost

    def roll_slots(self):
        """Display four 'slots' each with a randomly selected colour in each slot"""
        colours = ['black', 'white', 'green', 'yellow']
        return [random.choice(colours), random.choice(colours), random.choice(colours), random.choice(colours)]

    def unique_set_win(self, slots):
        """If each slot has a different colour then the machine should pay out half the current money in the machine"""
        if len(set(slots)) == len(slots):
            return True

    def adjacent_win(self, slots):
        """If two or more adjacent slots contain the same colour then the machine should pay out a prize of 5 times the cost of a single play"""
        for index in range(len(slots) - 1):
            if slots[index] == slots[index + 1]:
                return True

    def determine_prize(self, result):
        """Check all win conditions from major to minor wins and return prize amount"""
        if self.unique_set_win(result):
            prize = int(self.machine_bank / 2)
            self.machine_bank -= prize
            return prize
        elif self.adjacent_win(result):
            if self.machine_bank < self.play_cost * 5:
                self._provide_play_credit(self.play_cost * 5)
                prize = self.machine_bank
                self.machine_bank = 0
                return prize
            else:
                return self.play_cost * 5
        else:
            return 0

    def _provide_play_credit(self, target_payout):
        """House always wins: rounds down owed cash for machine credit"""
        credit = (target_payout - self.machine_bank) / self.play_cost
        self.play_credit += credit

# ########### GAME START ########### #

if __name__ == '__main__':

    machine1 = Machine(2222, 50)
    machine2 = Machine(1111, 10)

    player = Player(100)

    player.gamble(machine1)

File: fruit-machine/fruit_machine.rb

class Player
  attr_accessor :wallet_fund, :play_credit

  def initialize(wallet_fund)
    @wallet_fund = wallet_fund
    @play_credit = 0
  end

  def gamble(machine)
    if (@wallet_fund - machine.play_cost) > 0
      @wallet_fund -= machine.play_cost
      machine.run
    end
  end
end

class Machine
  attr_reader :slots, :play_cost, :play_credit
  attr_accessor :machine_bank

  def initialize(machine_bank, play_cost)
    @machine_bank = machine_bank
    @play_cost = play_cost
    @play_credit = 0
  end

  def run
    roll = roll_slots
    puts roll.to_s
    prize = determine_prize(roll)
    puts "WINNINGS: #{prize}"
  end

  def roll_slots
    colours = ['black', 'white', 'green', 'yellow']
    [colours.sample, colours.sample, colours.sample, colours.sample]
  end

  def unique_set_win?(slots)
    slots.uniq.length == slots.length ? true : false
  end

  def adjacent_win?(slots)
    slots.each_with_index do |slot, index|
      next_slot = slots[index + 1]
      if next_slot.nil?
        return false
      elsif slot == next_slot
        return true
      end
    end
  end

  def determine_prize(result)
    if unique_set_win?(result)
      prize = @machine_bank / 2
      @machine_bank -= prize
      return prize
    elsif adjacent_win?(result)
      if @machine_bank < @play_cost * 5
        provide_play_credit(@play_cost * 5)
        prize = @machine_bank
        @machine_bank = 0
        return prize
      else
        @machine_bank -= @play_cost * 5
        return @play_cost * 5
      end
    else
      return 0
    end
  end

  def provide_play_credit(target_payout)
    credit = (target_payout - @machine_bank) / @play_cost
    @play_credit += credit
  end
end

# ########### GAME START ########### #

if __FILE__ == $PROGRAM_NAME
  machine = Machine.new(2222, 50)
  player = Player.new(100)

  player.gamble(machine)
end
